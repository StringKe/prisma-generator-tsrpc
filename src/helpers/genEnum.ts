import { DMMF } from '@prisma/generator-helper'
import * as ts from 'typescript'

export const genEnum = ({
  name,
  values,
  documentation,
}: DMMF.DatamodelEnum) => {
  const enumValues = values.map((value) => {
    return ts.factory.createPropertySignature(
      undefined,
      ts.factory.createIdentifier(value.name),
      undefined,
      ts.factory.createLiteralTypeNode(
        ts.factory.createStringLiteral(value.name),
      ),
    )
  })

  const enumDeclaration = ts.factory.createVariableStatement(
    [ts.factory.createModifier(ts.SyntaxKind.ExportKeyword)],
    ts.factory.createVariableDeclarationList(
      [
        ts.factory.createVariableDeclaration(
          ts.factory.createIdentifier(name),
          undefined,
          ts.factory.createTypeLiteralNode(enumValues),
        ),
      ],
      ts.NodeFlags.Const,
    ),
  )

  ts.addSyntheticLeadingComment(
    enumDeclaration,
    ts.SyntaxKind.MultiLineCommentTrivia,
    ` Enum ${documentation} `,
    true,
  )

  const printer = ts.createPrinter()
  const enumStr = printer.printNode(
    ts.EmitHint.Unspecified,
    enumDeclaration,
    ts.createSourceFile('', '', ts.ScriptTarget.Latest),
  )

  const typeDeclaration = ts.factory.createTypeAliasDeclaration(
    [ts.factory.createModifier(ts.SyntaxKind.ExportKeyword)],
    ts.factory.createIdentifier(name),
    [],
    ts.factory.createIndexedAccessTypeNode(
      ts.factory.createParenthesizedType(
        ts.factory.createTypeQueryNode(ts.factory.createIdentifier(name)),
      ),
      ts.factory.createTypeOperatorNode(
        ts.SyntaxKind.KeyOfKeyword,
        ts.factory.createTypeQueryNode(ts.factory.createIdentifier(name)),
      ),
    ),
  )

  const typeStr = printer.printNode(
    ts.EmitHint.Unspecified,
    typeDeclaration,
    ts.createSourceFile('', '', ts.ScriptTarget.Latest),
  )

  return `${enumStr}\n\n${typeStr}\n\n`
}
