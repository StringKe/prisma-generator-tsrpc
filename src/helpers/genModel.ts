import { DMMF } from '@prisma/generator-helper'
import * as ts from 'typescript'

export const genModel = (
  { name, fields, documentation }: DMMF.Model,
  includeRelation = true,
) => {
  fields = includeRelation
    ? fields
    : fields.filter((field) => {
        return (
          field.relationName === undefined || field.relationName.length === 0
        )
      })

  const fieldDeclarations: ts.PropertySignature[] = fields.map((field) => {
    const fieldName = field.name

    let fieldType = 'any'
    if (field.kind === 'enum') {
      fieldType = field.type
    } else if (field.kind === 'object') {
      fieldType = field.type
    } else if (field.kind === 'scalar') {
      if (['Int', 'Float', 'Decimal', 'BigInt'].includes(field.type)) {
        fieldType = 'number'
      } else if (field.type === 'Boolean') {
        fieldType = 'boolean'
      } else if (field.type === 'DateTime') {
        fieldType = 'Date'
      } else if (field.type === 'String') {
        fieldType = 'string'
      } else if (field.type === 'Bytes') {
        fieldType = 'Buffer'
      } else if (field.type === 'Json') {
        fieldType = 'any'
      }
    } else if (field.kind === 'unsupported') {
      fieldType = 'any'
    }

    const isOptional = !field.isRequired
    const isList = field.isList

    const nodes = [
      isList
        ? ts.factory.createArrayTypeNode(
            ts.factory.createTypeReferenceNode(fieldType),
          )
        : ts.factory.createTypeReferenceNode(fieldType),
      isOptional ? ts.factory.createNull() : undefined,
    ].filter(Boolean) as ts.TypeNode[]

    const propertySignature = ts.factory.createPropertySignature(
      undefined,
      fieldName,
      undefined,
      ts.factory.createUnionTypeNode(nodes),
    )

    if (field.documentation) {
      ts.addSyntheticLeadingComment(
        propertySignature,
        ts.SyntaxKind.MultiLineCommentTrivia,
        ` ${field.documentation} `,
        true,
      )
    }
    return propertySignature
  })

  const interfaceDeclaration = ts.factory.createTypeAliasDeclaration(
    [ts.factory.createModifier(ts.SyntaxKind.ExportKeyword)],
    ts.factory.createIdentifier(!includeRelation ? `${name}` : `${name}Full`),
    [],
    ts.factory.createTypeLiteralNode(fieldDeclarations),
  )

  ts.addSyntheticLeadingComment(
    interfaceDeclaration,
    ts.SyntaxKind.MultiLineCommentTrivia,
    ` Model ${documentation} `,
    true,
  )

  const printer = ts.createPrinter()
  return printer.printNode(
    ts.EmitHint.Unspecified,
    interfaceDeclaration,
    ts.createSourceFile('', '', ts.ScriptTarget.Latest),
  )
}
