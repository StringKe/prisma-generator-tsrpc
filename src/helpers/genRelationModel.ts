import { DMMF } from '@prisma/generator-helper'
import * as ts from 'typescript'

function fcUpperFirst(str: string) {
  return str[0].toUpperCase() + str.slice(1)
}

export const genRelationModel = ({
  name,
  fields,
  documentation,
}: DMMF.Model) => {
  const fieldDeclarations: ts.PropertySignature[] = fields
    .filter((field) => field.kind === 'object')
    .map((field) => {
      const fieldName = field.name

      let fieldType = field.type

      const isOptional = !field.isRequired
      const isList = field.isList

      const nodes = [
        isList
          ? ts.factory.createArrayTypeNode(
              ts.factory.createTypeReferenceNode(fieldType),
            )
          : ts.factory.createTypeReferenceNode(fieldType),
        isOptional ? ts.factory.createNull() : undefined,
      ].filter(Boolean) as ts.TypeNode[]

      const propertySignature = ts.factory.createPropertySignature(
        undefined,
        fieldName,
        isOptional
          ? ts.factory.createToken(ts.SyntaxKind.QuestionToken)
          : undefined,
        ts.factory.createUnionTypeNode(nodes),
      )

      if (field.documentation) {
        ts.addSyntheticLeadingComment(
          propertySignature,
          ts.SyntaxKind.MultiLineCommentTrivia,
          ` ${field.documentation} `,
          true,
        )
      }
      return propertySignature
    })

  const strs = []

  const interfaceDeclaration = ts.factory.createTypeAliasDeclaration(
    [ts.factory.createModifier(ts.SyntaxKind.ExportKeyword)],
    ts.factory.createIdentifier(`${name}Relation`),
    [],
    ts.factory.createTypeLiteralNode(fieldDeclarations),
  )

  ts.addSyntheticLeadingComment(
    interfaceDeclaration,
    ts.SyntaxKind.MultiLineCommentTrivia,
    ` Model ${name} Relation All `,
    true,
  )

  const printer = ts.createPrinter()
  const str = printer.printNode(
    ts.EmitHint.Unspecified,
    interfaceDeclaration,
    ts.createSourceFile('', '', ts.ScriptTarget.Latest),
  )
  strs.push(str)

  for (const field of fieldDeclarations) {
    const filedName = field.name as ts.Identifier

    const interfaceDeclaration = ts.factory.createTypeAliasDeclaration(
      [ts.factory.createModifier(ts.SyntaxKind.ExportKeyword)],
      ts.factory.createIdentifier(
        `${name}${fcUpperFirst(filedName.escapedText as string)}Relation`,
      ),
      [],
      ts.factory.createTypeLiteralNode([field]),
    )

    ts.addSyntheticLeadingComment(
      interfaceDeclaration,
      ts.SyntaxKind.MultiLineCommentTrivia,
      ` Model ${name} Relation ${fcUpperFirst(
        filedName.escapedText as string,
      )} `,
      true,
    )

    const printer = ts.createPrinter()
    const str = printer.printNode(
      ts.EmitHint.Unspecified,
      interfaceDeclaration,
      ts.createSourceFile('', '', ts.ScriptTarget.Latest),
    )
    strs.push(str)
  }

  return strs.join('\n\n')
}
