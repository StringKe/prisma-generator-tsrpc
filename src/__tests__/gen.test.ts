import { genEnum } from '../helpers/genEnum'
import { getSampleDMMF } from './__fixtures__/getSampleDMMF'
import { genModel } from '../helpers/genModel'
import * as fs from 'fs'

test('enum generation', async () => {
  const sampleDMMF = await getSampleDMMF()
  fs.writeFileSync(
    './src/__tests__/__fixtures__/sampleDMMF.json',
    JSON.stringify(sampleDMMF, null, 2),
  )
  const enums = sampleDMMF.datamodel.enums
    .map((enumInfo) => {
      return genEnum(enumInfo)
    })
    .join('\n')
  expect(enums).toMatchSnapshot()
})

test('model generation', async () => {
  const sampleDMMF = await getSampleDMMF()
  const models = sampleDMMF.datamodel.models
    .map((model) => {
      return genModel(model as never)
    })
    .join('\n')
  expect(models).toMatchSnapshot()
})
