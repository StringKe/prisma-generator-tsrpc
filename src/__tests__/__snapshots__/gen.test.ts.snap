// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`enum generation 1`] = `
"export enum NOTIFICATION_TYPE {
    NEW_POSTS = "newPosts",
    NEW_COMMENTS = "newComments",
    NEW_FOLLOWERS = "newFollowers",
    REPLY = "reply",
    HEART_ON_POST = "heartOnPost",
    HEART_ON_COMMENT = "heartOnComment",
    HEART_ON_REPLY = "heartOnReply"
}
export enum LANGUAGE {
    TYPESCRIPT = "Typescript",
    JAVASCRIPT = "Javascript",
    RUST = "Rust",
    GO = "Go",
    PYTHON = "Python",
    CPP = "Cpp"
}
export enum STATE {
    ACTIVE = "Active",
    PENDING = "Pending",
    BANNED = "Banned"
}"
`;

exports[`model generation 1`] = `
"export interface User {
    id: number;
    email: string;
    name?: string;
}
export interface Product {
    id: number;
    title: string;
    desc: string;
    images: any;
    isShown?: boolean;
    stock?: number;
    type: STATE;
    averageRating?: number;
    categoryId: number;
    companyId: number;
    category: Category;
    company: Company;
    createdAt: Date;
    updatedAt: Date;
}
export interface Category {
    id: number;
    products: Product[];
}
export interface Company {
    id: number;
    name: string;
    totalIncome: number;
    lat: number;
    lng: number;
    by: Buffer;
    products: Product[];
}"
`;
